kotlinx-coroutines-test/common/src/TestScope.kt

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.test
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+
+/**
+ * A coroutine scope that for launching test coroutines.
+ *
+ * The scope provides the following functionality:
+ * * The [coroutineContext] includes a [coroutine dispatcher][TestDispatcher] that supports delay-skipping, using
+ *   a [TestCoroutineScheduler] for orchestrating the virtual time.
+ *   This scheduler is also available via the [testScheduler] property, and some helper extension
+ *   methods are defined to more conveniently interact with it: see [TestScope.currentTime], [TestScope.runCurrent],
+ *   [TestScope.advanceTimeBy], and [TestScope.advanceUntilIdle].
+ * * When inside [runTest], uncaught exceptions from the child coroutines of this scope will be reported at the end of
+ *   the test.
+ *   It is invalid for child coroutines to throw uncaught exceptions when outside the call to [TestScope.runTest]:
+ *   the only guarantee in this case is the best effort to deliver the exception.
+ *
+ * The usual way to access a [TestScope] is to call [runTest], but it can also be constructed manually, in order to
+ * use it to initialize the components that participate in the test.
+ *
+ * #### Differences from [TestCoroutineScope]
+ *
+ * * This doesn't provide an equivalent of [TestCoroutineScope.cleanupTestCoroutines], and so can't be used as a
+ *   standalone mechanism for writing tests: it does require that [runTest] is eventually called.
+ *   The reason for this is that a proper cleanup procedure that supports using non-test dispatchers and arbitrary
+ *   coroutine suspensions would be equivalent to [runTest], but would also be more error-prone, due to the potential
+ *   for forgetting to perform the cleanup.
+ * * [TestCoroutineScope.advanceTimeBy] also calls [TestCoroutineScheduler.runCurrent] after advancing the virtual time.
+ * * No support for dispatcher pausing, like [DelayController] allows. [TestCoroutineDispatcher], which supported
+ *   pausing, is deprecated; now, instead of pausing a dispatcher, one can use [withContext] to run a dispatcher that's
+ *   paused by default, like [StandardTestDispatcher].
+ * * No access to the list of unhandled exceptions.
+ */
+@ExperimentalCoroutinesApi
+public sealed interface TestScope : CoroutineScope {
+    /**
+     * The delay-skipping scheduler used by the test dispatchers running the code in this scope.
+     */
+    @ExperimentalCoroutinesApi
+    public val testScheduler: TestCoroutineScheduler
```

Comment:
Shouldn't it also provide typed access to `TestDispatcher`?

The benefits are the same as for creating `TestScope` standalone -- capability to configure entities that depend on either scope or dispatcher.

It's possible to argue that one can use `coroutineContext[CoroutineDispatcher]!!`, but it is both verbose and required `ExperimentalStdlibApi`, while having an explicit property with type `TestDispatcher` clearly indicates that scope is always bound to the instance of test dispatcher

Reply:
> `coroutineContext[CoroutineDispatcher]!!`

Or `StandardTestDispatcher(testCoroutineScheduler)`. I didn't want to prioritize specific dispatchers.

I believe that the reason to pass the test dispatcher somewhere explicitly is to ensure that a coroutine launched somewhere operates using the source of virtual time (from the sound of it I already get the feeling that it's the source of virtual time that should be passed), not a specific dispatching strategy: if it was the question of the strategy, then this coroutine would just be launched in this scope and inherit the dispatcher automatically.

From this point of view, both `StandardTestDispatcher(testCoroutineScheduler)` and `UnconfinedTestDispatcher(testCoroutineScheduler)` are valid, symmetric choices. However, introducing a `testDispatcher` field promotes one dispatching strategy (the one used for the scope) over the other.

Finally, if one does for some reason need to make sure that the scope and the other parts of the system use the same dispatcher, then this (I think, very specific) requirement can be fulfilled using
```kotlin
val dispatcher = StandardTestDispatcher()
val scope = TestScope(dispatcher)
```

Maybe I'm missing something and the need to pass the specific dispatcher would be much more frequent than I think?